File: v2/packages/core/src/render.ts
/**
 * Canonical JSON debug rendering (SPEC.md §11).
 *
 * Non-contractual for the wire; contractual for golden vectors. Rendering
 * rules: frames in wire order with END omitted, field order per the SPEC
 * field tables, absent opt() fields omitted, binary fields as standard
 * base64, enums by their spec names, i64 as JSON numbers, json-typed
 * fields embedded as parsed JSON.
 */
import type {
  CommitChange,
  PushOperation,
  PushOperationResult,
  RequestFrame,
  ResponseFrame,
  SyncMessage,
} from './message';
import { decodeMessage } from './message';
import type { RowColumn, RowValue } from './row-codec';
import type { RowsSegment } from './segment';
import { decodeRowsSegment, ROWS_SEGMENT_FORMAT_VERSION } from './segment';

export type JsonValue =
  | null
  | boolean
  | number
  | string
  | JsonValue[]
  | { [key: string]: JsonValue };

const BASE64_ALPHABET =
  'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';

/** Standard base64 (§11.1 rule 4). Dependency- and runtime-agnostic. */
export function bytesToBase64(bytes: Uint8Array): string {
  let out = '';
  let i = 0;
  for (; i + 2 < bytes.length; i += 3) {
    const chunk =
      ((bytes[i] ?? 0) << 16) |
      ((bytes[i + 1] ?? 0) << 8) |
      (bytes[i + 2] ?? 0);
    out +=
      BASE64_ALPHABET.charAt((chunk >> 18) & 63) +
      BASE64_ALPHABET.charAt((chunk >> 12) & 63) +
      BASE64_ALPHABET.charAt((chunk >> 6) & 63) +
      BASE64_ALPHABET.charAt(chunk & 63);
  }
  const rest = bytes.length - i;
  if (rest === 1) {
    const chunk = (bytes[i] ?? 0) << 16;
    out +=
      BASE64_ALPHABET.charAt((chunk >> 18) & 63) +
      BASE64_ALPHABET.charAt((chunk >> 12) & 63) +
      '==';
  } else if (rest === 2) {
    const chunk = ((bytes[i] ?? 0) << 16) | ((bytes[i + 1] ?? 0) << 8);
    out +=
      BASE64_ALPHABET.charAt((chunk >> 18) & 63) +
      BASE64_ALPHABET.charAt((chunk >> 12) & 63) +
      BASE64_ALPHABET.charAt((chunk >> 6) & 63) +
      '=';
  }
  return out;
}

function parseJson(raw: string): JsonValue {
  return JSON.parse(raw) as JsonValue;
}

function renderScopeLists(
  scopes: Readonly<Record<string, string[]>>,
): JsonValue {
  return Object.fromEntries(
    Object.entries(scopes).map(([key, values]) => [key, [...values]]),
  );
}

function renderStringMap(map: Readonly<Record<string, string>>): JsonValue {
  return { ...map };
}

function renderOperation(operation: PushOperation): JsonValue {
  return {
    table: operation.table,
    rowId: operation.rowId,
    op: operation.op,
    ...(operation.baseVersion !== undefined
      ? { baseVersion: operation.baseVersion }
      : {}),
    ...(operation.payload !== undefined
      ? { payload: bytesToBase64(operation.payload) }
      : {}),
  };
}

function renderResult(result: PushOperationResult): JsonValue {
  if (result.status === 'applied') {
    return { opIndex: result.opIndex, status: 'applied' };
  }
  if (result.status === 'conflict') {
    return {
      opIndex: result.opIndex,
      status: 'conflict',
      code: result.code,
      message: result.message,
      serverVersion: result.serverVersion,
      serverRow: bytesToBase64(result.serverRow),
    };
  }
  return {
    opIndex: result.opIndex,
    status: 'error',
    code: result.code,
    message: result.message,
    retryable: result.retryable,
  };
}

function renderChange(change: CommitChange): JsonValue {
  return {
    tableIndex: change.tableIndex,
    rowId: change.rowId,
    op: change.op,
    ...(change.rowVersion !== undefined
      ? { rowVersion: change.rowVersion }
      : {}),
    scopes: renderStringMap(change.scopes),
    ...(change.row !== undefined ? { row: bytesToBase64(change.row) } : {}),
  };
}

function renderFrame(frame: RequestFrame | ResponseFrame): JsonValue {
  switch (frame.type) {
    case 'REQ_HEADER':
      return {
        type: 'REQ_HEADER',
        clientId: frame.clientId,
        schemaVersion: frame.schemaVersion,
      };
    case 'PUSH_COMMIT':
      return {
        type: 'PUSH_COMMIT',
        clientCommitId: frame.clientCommitId,
        operations: frame.operations.map(renderOperation),
      };
    case 'PULL_HEADER':
      return {
        type: 'PULL_HEADER',
        limitCommits: frame.limitCommits,
        limitSnapshotRows: frame.limitSnapshotRows,
        maxSnapshotPages: frame.maxSnapshotPages,
        accept: frame.accept,
      };
    case 'SUBSCRIPTION':
      return {
        type: 'SUBSCRIPTION',
        id: frame.id,
        table: frame.table,
        scopes: renderScopeLists(frame.scopes),
        ...(frame.params !== undefined
          ? { params: parseJson(frame.params) }
          : {}),
        cursor: frame.cursor,
        ...(frame.bootstrapState !== undefined
          ? { bootstrapState: parseJson(frame.bootstrapState) }
          : {}),
      };
    case 'RESP_HEADER':
      return {
        type: 'RESP_HEADER',
        ...(frame.requiredSchemaVersion !== undefined
          ? { requiredSchemaVersion: frame.requiredSchemaVersion }
          : {}),
        ...(frame.latestSchemaVersion !== undefined
          ? { latestSchemaVersion: frame.latestSchemaVersion }
          : {}),
      };
    case 'PUSH_RESULT':
      return {
        type: 'PUSH_RESULT',
        clientCommitId: frame.clientCommitId,
        status: frame.status,
        ...(frame.commitSeq !== undefined
          ? { commitSeq: frame.commitSeq }
          : {}),
        results: frame.results.map(renderResult),
      };
    case 'SUB_START':
      return {
        type: 'SUB_START',
        id: frame.id,
        status: frame.status,
        reasonCode: frame.reasonCode,
        effectiveScopes: renderScopeLists(frame.effectiveScopes),
        bootstrap: frame.bootstrap,
      };
    case 'COMMIT':
      return {
        type: 'COMMIT',
        commitSeq: frame.commitSeq,
        createdAtMs: frame.createdAtMs,
        actorId: frame.actorId,
        tables: [...frame.tables],
        changes: frame.changes.map(renderChange),
      };
    case 'SEGMENT_REF':
      return {
        type: 'SEGMENT_REF',
        segmentId: frame.segmentId,
        mediaType: frame.mediaType,
        table: frame.table,
        byteLength: frame.byteLength,
        rowCount: frame.rowCount,
        asOfCommitSeq: frame.asOfCommitSeq,
        scopeDigest: frame.scopeDigest,
        ...(frame.rowCursor !== undefined
          ? { rowCursor: frame.rowCursor }
          : {}),
        ...(frame.nextRowCursor !== undefined
          ? { nextRowCursor: frame.nextRowCursor }
          : {}),
        ...(frame.url !== undefined ? { url: frame.url } : {}),
        ...(frame.urlExpiresAtMs !== undefined
          ? { urlExpiresAtMs: frame.urlExpiresAtMs }
          : {}),
      };
    case 'SEGMENT_INLINE':
      return {
        type: 'SEGMENT_INLINE',
        payload: bytesToBase64(frame.payload),
      };
    case 'SUB_END':
      return {
        type: 'SUB_END',
        nextCursor: frame.nextCursor,
        ...(frame.bootstrapState !== undefined
          ? { bootstrapState: parseJson(frame.bootstrapState) }
          : {}),
      };
    case 'ERROR':
      return {
        type: 'ERROR',
        code: frame.code,
        message: frame.message,
        category: frame.category,
        retryable: frame.retryable,
        recommendedAction: frame.recommendedAction,
        ...(frame.details !== undefined
          ? { details: parseJson(frame.details) }
          : {}),
      };
    case 'UNKNOWN':
      return {
        type: 'UNKNOWN',
        frameType: frame.frameType,
        payload: bytesToBase64(frame.payload),
      };
  }
}

export function renderMessageValue(message: SyncMessage): JsonValue {
  return {
    magic: 'SSP2',
    wireVersion: message.wireVersion,
    msgKind: message.msgKind,
    frames: message.frames.map((frame: RequestFrame | ResponseFrame) =>
      renderFrame(frame),
    ),
  };
}

/** `render(bytes) → json` for SSP2 messages (§11). */
export function renderMessage(bytes: Uint8Array): JsonValue {
  return renderMessageValue(decodeMessage(bytes));
}

function renderRowValue(column: RowColumn, value: RowValue): JsonValue {
  if (value === null) return null;
  if (value instanceof Uint8Array) return bytesToBase64(value);
  if (
    (column.type === 'json' || column.type === 'blob_ref') &&
    typeof value === 'string'
  ) {
    // §11: json and blob_ref (tag 7) render as embedded parsed JSON.
    return parseJson(value);
  }
  return value;
}

export function renderRowsSegmentValue(segment: RowsSegment): JsonValue {
  return {
    magic: 'SSG2',
    formatVersion: ROWS_SEGMENT_FORMAT_VERSION,
    table: segment.table,
    schemaVersion: segment.schemaVersion,
    columns: segment.columns.map((column) => ({
      name: column.name,
      type: column.type,
      nullable: column.nullable,
    })),
    blocks: segment.blocks.map((block) =>
      block.map((row) => ({
        serverVersion: row.serverVersion,
        values: Object.fromEntries(
          segment.columns.map((column, i) => [
            column.name,
            renderRowValue(column, row.values[i] ?? null),
          ]),
        ),
      })),
    ),
  };
}

/** `render(bytes) → json` for standalone SSG2 rows segments (§11 rule 8). */
export function renderRowsSegment(bytes: Uint8Array): JsonValue {
  return renderRowsSegmentValue(decodeRowsSegment(bytes));
}

File: rust/crates/ssp2/src/render.rs
//! Canonical JSON debug rendering (SPEC.md §11.1). Non-contractual for the
//! wire; contractual for golden vectors — rendered values must deep-equal the
//! committed `.json` files when parsed as JSON.

use serde_json::{Map, Value};

use crate::model::{Frame, Message, MsgKind, OpResult};
use crate::segment::{ColumnValue, RowsSegment};
use crate::util::base64;

fn obj(entries: Vec<(&str, Value)>) -> Value {
    let mut map = Map::new();
    for (k, v) in entries {
        map.insert(k.to_string(), v);
    }
    Value::Object(map)
}

fn scope_map_value(scopes: &[(String, Vec<String>)]) -> Value {
    let mut map = Map::new();
    for (k, vs) in scopes {
        map.insert(
            k.clone(),
            Value::Array(vs.iter().map(|v| Value::String(v.clone())).collect()),
        );
    }
    Value::Object(map)
}

/// Render a decoded message per §11.1 rule 1: one JSON document, frames in
/// wire order, `END` omitted.
pub fn render_message(msg: &Message) -> Value {
    obj(vec![
        ("magic", Value::from("SSP2")),
        ("wireVersion", Value::from(1)),
        (
            "msgKind",
            Value::from(match msg.msg_kind {
                MsgKind::Request => "request",
                MsgKind::Response => "response",
            }),
        ),
        (
            "frames",
            Value::Array(msg.frames.iter().map(render_frame).collect()),
        ),
    ])
}

fn render_frame(frame: &Frame) -> Value {
    // Rule 2: field order follows the SPEC field tables (JSON object key
    // order is irrelevant for deep-equality, but we keep it anyway).
    // Rule 3: absent opt() fields are omitted, never null.
    let mut entries: Vec<(&str, Value)> = Vec::new();
    match frame {
        Frame::ReqHeader {
            client_id,
            schema_version,
        } => {
            entries.push(("type", Value::from("REQ_HEADER")));
            entries.push(("clientId", Value::from(client_id.clone())));
            entries.push(("schemaVersion", Value::from(*schema_version)));
        }
        Frame::PushCommit {
            client_commit_id,
            operations,
        } => {
            entries.push(("type", Value::from("PUSH_COMMIT")));
            entries.push(("clientCommitId", Value::from(client_commit_id.clone())));
            let ops = operations
                .iter()
                .map(|op| {
                    let mut e: Vec<(&str, Value)> = vec![
                        ("table", Value::from(op.table.clone())),
                        ("rowId", Value::from(op.row_id.clone())),
                        ("op", Value::from(op.op.name())),
                    ];
                    if let Some(v) = op.base_version {
                        e.push(("baseVersion", Value::from(v)));
                    }
                    if let Some(p) = &op.payload {
                        e.push(("payload", Value::from(base64(p))));
                    }
                    obj(e)
                })
                .collect();
            entries.push(("operations", Value::Array(ops)));
        }
        Frame::PullHeader {
            limit_commits,
            limit_snapshot_rows,
            max_snapshot_pages,
            accept,
        } => {
            entries.push(("type", Value::from("PULL_HEADER")));
            entries.push(("limitCommits", Value::from(*limit_commits)));
            entries.push(("limitSnapshotRows", Value::from(*limit_snapshot_rows)));
            entries.push(("maxSnapshotPages", Value::from(*max_snapshot_pages)));
            // Rule 5: bitmask fields are not enums; render numerically.
            entries.push(("accept", Value::from(*accept)));
        }
        Frame::Subscription {
            id,
            table,
            scopes,
            params,
            cursor,
            bootstrap_state,
        } => {
            entries.push(("type", Value::from("SUBSCRIPTION")));
            entries.push(("id", Value::from(id.clone())));
            entries.push(("table", Value::from(table.clone())));
            entries.push(("scopes", scope_map_value(scopes)));
            if let Some(p) = params {
                // Rule 7: json-typed fields embed as parsed JSON.
                entries.push(("params", p.parse()));
            }
            entries.push(("cursor", Value::from(*cursor)));
            if let Some(b) = bootstrap_state {
                entries.push(("bootstrapState", b.parse()));
            }
        }
        Frame::RespHeader {
            required_schema_version,
            latest_schema_version,
        } => {
            entries.push(("type", Value::from("RESP_HEADER")));
            if let Some(v) = required_schema_version {
                entries.push(("requiredSchemaVersion", Value::from(*v)));
            }
            if let Some(v) = latest_schema_version {
                entries.push(("latestSchemaVersion", Value::from(*v)));
            }
        }
        Frame::Lease {
            lease_id,
            expires_at_ms,
        } => {
            entries.push(("type", Value::from("LEASE")));
            entries.push(("leaseId", Value::from(lease_id.clone())));
            entries.push(("expiresAtMs", Value::from(*expires_at_ms)));
        }
        Frame::PushResult {
            client_commit_id,
            status,
            commit_seq,
            results,
        } => {
            entries.push(("type", Value::from("PUSH_RESULT")));
            entries.push(("clientCommitId", Value::from(client_commit_id.clone())));
            entries.push(("status", Value::from(status.name())));
            if let Some(v) = commit_seq {
                entries.push(("commitSeq", Value::from(*v)));
            }
            entries.push((
                "results",
                Value::Array(results.iter().map(render_result).collect()),
            ));
        }
        Frame::PushResultDetails {
            client_commit_id,
            entries: details_entries,
        } => {
            entries.push(("type", Value::from("PUSH_RESULT_DETAILS")));
            entries.push(("clientCommitId", Value::from(client_commit_id.clone())));
            entries.push((
                "entries",
                Value::Array(
                    details_entries
                        .iter()
                        .map(|entry| {
                            obj(vec![
                                ("opIndex", Value::from(entry.op_index)),
                                ("details", entry.details.parse()),
                            ])
                        })
                        .collect(),
                ),
            ));
        }
        Frame::SubStart {
            id,
            status,
            reason_code,
            effective_scopes,
            bootstrap,
        } => {
            entries.push(("type", Value::from("SUB_START")));
            entries.push(("id", Value::from(id.clone())));
            entries.push(("status", Value::from(status.name())));
            entries.push(("reasonCode", Value::from(reason_code.clone())));
            entries.push(("effectiveScopes", scope_map_value(effective_scopes)));
            entries.push(("bootstrap", Value::from(*bootstrap)));
        }
        Frame::Commit {
            commit_seq,
            created_at_ms,
            actor_id,
            tables,
            changes,
        } => {
            entries.push(("type", Value::from("COMMIT")));
            entries.push(("commitSeq", Value::from(*commit_seq)));
            entries.push(("createdAtMs", Value::from(*created_at_ms)));
            entries.push(("actorId", Value::from(actor_id.clone())));
            entries.push((
                "tables",
                Value::Array(tables.iter().map(|t| Value::from(t.clone())).collect()),
            ));
            let rendered_changes = changes
                .iter()
                .map(|c| {
                    let mut e: Vec<(&str, Value)> = vec![
                        ("tableIndex", Value::from(c.table_index)),
                        ("rowId", Value::from(c.row_id.clone())),
                        ("op", Value::from(c.op.name())),
                    ];
                    if let Some(v) = c.row_version {
                        e.push(("rowVersion", Value::from(v)));
                    }
                    let mut scopes = Map::new();
                    for (k, v) in &c.scopes {
                        scopes.insert(k.clone(), Value::from(v.clone()));
                    }
                    e.push(("scopes", Value::Object(scopes)));
                    if let Some(row) = &c.row {
                        // Rule 4: binary fields render as base64.
                        e.push(("row", Value::from(base64(row))));
                    }
                    obj(e)
                })
                .collect();
            entries.push(("changes", Value::Array(rendered_changes)));
        }
        Frame::SegmentRef {
            segment_id,
            media_type,
            table,
            byte_length,
            row_count,
            as_of_commit_seq,
            scope_digest,
            row_cursor,
            next_row_cursor,
            url,
            url_expires_at_ms,
        } => {
            entries.push(("type", Value::from("SEGMENT_REF")));
            entries.push(("segmentId", Value::from(segment_id.clone())));
            entries.push(("mediaType", Value::from(media_type.name())));
            entries.push(("table", Value::from(table.clone())));
            entries.push(("byteLength", Value::from(*byte_length)));
            entries.push(("rowCount", Value::from(*row_count)));
            entries.push(("asOfCommitSeq", Value::from(*as_of_commit_seq)));
            entries.push(("scopeDigest", Value::from(scope_digest.clone())));
            if let Some(v) = row_cursor {
                entries.push(("rowCursor", Value::from(v.clone())));
            }
            if let Some(v) = next_row_cursor {
                entries.push(("nextRowCursor", Value::from(v.clone())));
            }
            if let Some(v) = url {
                entries.push(("url", Value::from(v.clone())));
            }
            if let Some(v) = url_expires_at_ms {
                entries.push(("urlExpiresAtMs", Value::from(*v)));
            }
        }
        Frame::SegmentInline { payload } => {
            // Rule 4: the whole payload renders as base64, never as a nested
            // rule-8 object.
            entries.push(("type", Value::from("SEGMENT_INLINE")));
            entries.push(("payload", Value::from(base64(payload))));
        }
        Frame::SubEnd {
            next_cursor,
            bootstrap_state,
        } => {
            entries.push(("type", Value::from("SUB_END")));
            entries.push(("nextCursor", Value::from(*next_cursor)));
            if let Some(b) = bootstrap_state {
                entries.push(("bootstrapState", b.parse()));
            }
        }
        Frame::Error {
            code,
            message,
            category,
            retryable,
            recommended_action,
            details,
        } => {
            entries.push(("type", Value::from("ERROR")));
            entries.push(("code", Value::from(code.clone())));
            entries.push(("message", Value::from(message.clone())));
            entries.push(("category", Value::from(category.clone())));
            entries.push(("retryable", Value::from(*retryable)));
            entries.push(("recommendedAction", Value::from(recommended_action.clone())));
            if let Some(d) = details {
                entries.push(("details", d.parse()));
            }
        }
        Frame::Unknown {
            frame_type,
            payload,
        } => {
            // Rule 9.
            entries.push(("type", Value::from("UNKNOWN")));
            entries.push(("frameType", Value::from(*frame_type)));
            entries.push(("payload", Value::from(base64(payload))));
        }
    }
    obj(entries)
}

fn render_result(result: &OpResult) -> Value {
    match result {
        OpResult::Applied { op_index } => obj(vec![
            ("opIndex", Value::from(*op_index)),
            ("status", Value::from("applied")),
        ]),
        OpResult::Conflict {
            op_index,
            code,
            message,
            server_version,
            server_row,
        } => obj(vec![
            ("opIndex", Value::from(*op_index)),
            ("status", Value::from("conflict")),
            ("code", Value::from(code.clone())),
            ("message", Value::from(message.clone())),
            ("serverVersion", Value::from(*server_version)),
            ("serverRow", Value::from(base64(server_row))),
        ]),
        OpResult::Error {
            op_index,
            code,
            message,
            retryable,
        } => obj(vec![
            ("opIndex", Value::from(*op_index)),
            ("status", Value::from("error")),
            ("code", Value::from(code.clone())),
            ("message", Value::from(message.clone())),
            ("retryable", Value::from(*retryable)),
        ]),
    }
}

/// Render a rows segment per §11.1 rule 8: row records as
/// `{"serverVersion":…,"values":{name→value}}`, NULLs as JSON `null`,
/// `bytes` as base64, `json` columns parsed.
pub fn render_rows_segment(seg: &RowsSegment) -> Value {
    let columns = seg
        .columns
        .iter()
        .map(|c| {
            obj(vec![
                ("name", Value::from(c.name.clone())),
                ("type", Value::from(c.ty.name())),
                ("nullable", Value::from(c.nullable)),
            ])
        })
        .collect();
    let blocks = seg
        .blocks
        .iter()
        .map(|block| {
            Value::Array(
                block
                    .iter()
                    .map(|row| {
                        let mut values = Map::new();
                        for (col, value) in seg.columns.iter().zip(row.values.iter()) {
                            values.insert(col.name.clone(), render_column_value(value));
                        }
                        obj(vec![
                            ("serverVersion", Value::from(row.server_version)),
                            ("values", Value::Object(values)),
                        ])
                    })
                    .collect(),
            )
        })
        .collect();
    obj(vec![
        ("magic", Value::from("SSG2")),
        ("formatVersion", Value::from(1)),
        ("table", Value::from(seg.table.clone())),
        ("schemaVersion", Value::from(seg.schema_version)),
        ("columns", Value::Array(columns)),
        ("blocks", Value::Array(blocks)),
    ])
}

fn render_column_value(value: &Option<ColumnValue>) -> Value {
    match value {
        None => Value::Null,
        Some(ColumnValue::String(s)) => Value::from(s.clone()),
        Some(ColumnValue::Integer(v)) => Value::from(*v),
        Some(ColumnValue::Float(v)) => {
            serde_json::Number::from_f64(*v).map_or(Value::Null, Value::Number)
        }
        Some(ColumnValue::Boolean(v)) => Value::from(*v),
        Some(ColumnValue::Json(j)) => j.parse(),
        Some(ColumnValue::Bytes(b)) => Value::from(base64(b)),
        // §11: blob_ref (tag 7) renders as embedded parsed JSON, like json.
        Some(ColumnValue::BlobRef(j)) => j.parse(),
        // §11: crdt (tag 8) renders as base64, like bytes.
        Some(ColumnValue::Crdt(b)) => Value::from(base64(b)),
    }
}
